src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -249,39 +252,47 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
+            var filterDataList = new SegmentedList<MatchResult>(items.Select(
+                item => helper.GetMatchResult(item, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
 
-            var builder = ImmutableArray.CreateBuilder<CompletionItem>();
-            FilterItems(helper, itemsWithPatternMatch, filterText, builder);
-            return builder.ToImmutable();
+            var builder = s_listOfMatchResultPool.Allocate();
+            try
+            {
+                FilterItems(helper, filterDataList, filterText, builder);
+                return builder.SelectAsArray(result => result.CompletionItem);
+            }
+            finally
+            {
+                // Don't call ClearAndFree, which resets the capacity to a default value.
+                builder.Clear();
+                s_listOfMatchResultPool.Free(builder);
+            }
         }
 
         internal virtual void FilterItems(
            Document document,
-           IReadOnlyList<(CompletionItem, PatternMatch?)> itemsWithPatternMatch,
+           IReadOnlyList<MatchResult> matchResults,
            string filterText,
-           IList<CompletionItem> builder)
+           IList<MatchResult> builder)
         {
 #pragma warning disable RS0030 // Do not used banned APIs
             // Default implementation just drops the pattern matches and builder, and calls the public overload of FilterItems instead for compatibility.
-            builder.AddRange(FilterItems(document, itemsWithPatternMatch.SelectAsArray(item => item.Item1), filterText));
+            var filteredItems = FilterItems(document, matchResults.SelectAsArray(item => item.CompletionItem), filterText);
 #pragma warning restore RS0030 // Do not used banned APIs
-        }
 
-        // The FilterItems method might need to handle a large list of items when import completion is enabled and filter text is
-        // very short, i.e. <= 1. Therefore, use pooled list to avoid repeated (potentially LOH) allocations.
-        private static readonly ObjectPool<List<(CompletionItem item, PatternMatch? match)>> s_listOfItemMatchPairPool = new(factory: () => new());
+            var helper = CompletionHelper.GetHelper(document);
+            builder.AddRange(filteredItems.Select(item => helper.GetMatchResult(item, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
+        }
 
         /// <summary>
         /// Determine among the provided items the best match w.r.t. the given filter text, 
         /// those returned would be considered equally good candidates for selection by controller.
         /// </summary>
         internal static void FilterItems(
             CompletionHelper completionHelper,
-            IReadOnlyList<(CompletionItem item, PatternMatch? match)> itemsWithPatternMatch,
+            IReadOnlyList<MatchResult> matchResults,
             string filterText,
-            IList<CompletionItem> builder)
+            IList<MatchResult> builder)
```

Comment:
Now we simply return a subset of `MatchResult`s passed in, this would provide more information across the API boundry and simply the implementation.

Reply:
consider updating comment